Description: Fix GnuTLS backend by postponing initialization after it daemonized.
 Backport upstream commits from version 15.2.
Origin: https://github.com/bareos/bareos/commit/9097aaeaefe904b40af602caddf5d9cd59959625
 https://github.com/bareos/bareos/commit/ecb539bc44c0224b378e6e9626b86ea718da5c2c

--- bareos-14.2.6.orig/src/dird/dird.c
+++ bareos-14.2.6/src/dird/dird.c
@@ -285,6 +285,13 @@ int main (int argc, char *argv[])
    my_config = new_config_parser();
    parse_dir_config(my_config, configfile, M_ERROR_TERM);
 
+   if (!test_config) {                /* we don't need to do this block in test mode */
+      if (background) {
+         daemon_start();
+         init_stack_dump();              /* grab new pid */
+      }
+   }
+
    if (init_crypto() != 0) {
       Jmsg((JCR *)NULL, M_ERROR_TERM, 0, _("Cryptography library initialization failed.\n"));
       goto bail_out;
@@ -296,10 +303,6 @@ int main (int argc, char *argv[])
    }
 
    if (!test_config) {                /* we don't need to do this block in test mode */
-      if (background) {
-         daemon_start();
-         init_stack_dump();              /* grab new pid */
-      }
       /* Create pid must come after we are a daemon -- so we have our final pid */
       create_pid_file(me->pid_directory, "bareos-dir",
                       get_first_port_host_order(me->DIRaddrs));
--- bareos-14.2.6.orig/src/filed/filed.c
+++ bareos-14.2.6/src/filed/filed.c
@@ -213,6 +213,11 @@ int main (int argc, char *argv[])
    my_config = new_config_parser();
    parse_fd_config(my_config, configfile, M_ERROR_TERM);
 
+   if (!foreground && !test_config) {
+      daemon_start();
+      init_stack_dump();              /* set new pid */
+   }
+
    if (init_crypto() != 0) {
       Emsg0(M_ERROR, 0, _("Cryptography library initialization failed.\n"));
       terminate_filed(1);
@@ -237,11 +242,6 @@ int main (int argc, char *argv[])
       terminate_filed(0);
    }
 
-   if (!foreground) {
-      daemon_start();
-      init_stack_dump();              /* set new pid */
-   }
-
    set_thread_concurrency(me->MaxConcurrentJobs + 10);
    lmgr_init_thread(); /* initialize the lockmanager stack */
 
--- bareos-14.2.6.orig/src/stored/stored.c
+++ bareos-14.2.6/src/stored/stored.c
@@ -219,6 +219,11 @@ int main (int argc, char *argv[])
    my_config = new_config_parser();
    parse_sd_config(my_config, configfile, M_ERROR_TERM);
 
+   if (!foreground && !test_config) {
+      daemon_start();                 /* become daemon */
+      init_stack_dump();              /* pick up new pid */
+   }
+
    if (init_crypto() != 0) {
       Jmsg((JCR *)NULL, M_ERROR_TERM, 0, _("Cryptography library initialization failed.\n"));
    }
@@ -235,11 +240,6 @@ int main (int argc, char *argv[])
 
    my_name_is(0, (char **)NULL, me->hdr.name);     /* Set our real name */
 
-   if (!foreground) {
-      daemon_start();                 /* become daemon */
-      init_stack_dump();              /* pick up new pid */
-   }
-
    create_pid_file(me->pid_directory, "bareos-sd",
                    get_first_port_host_order(me->SDaddrs));
    read_state_file(me->working_directory, "bareos-sd",
